File: platform/android/lynx_android/src/main/java/com/lynx/tasm/behavior/ui/utils/PlatformLength.java
// Copyright 2023 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

package com.lynx.tasm.behavior.ui.utils;

import androidx.annotation.Nullable;
import com.lynx.react.bridge.Dynamic;
import com.lynx.react.bridge.ReadableArray;
import com.lynx.tasm.behavior.StyleConstants;
import com.lynx.tasm.utils.FloatUtils;

/**
 * PlatformLength has unit and value.
 * The value can be an array with a sub-length or a float value.
 */
public class PlatformLength {
  private final int mType;
  private float mValue;
  // [value, unit, [value, unit, ...], unit, ...]
  private ReadableArray mArray;

  public PlatformLength(Dynamic value, int type) {
    if (type == StyleConstants.PLATFORM_LENGTH_UNIT_CALC) {
      mArray = value.asArray();
    } else if (type == StyleConstants.PLATFORM_LENGTH_UNIT_NUMBER
        || type == StyleConstants.PLATFORM_LENGTH_UNIT_PERCENT) {
      mValue = (float) value.asDouble();
    }
    mType = type;
  }

  public PlatformLength(float value, int type) {
    mValue = value;
    mType = type;
  }

  public float asNumber() {
    return mType == StyleConstants.PLATFORM_LENGTH_UNIT_NUMBER ? mValue : 0;
  }

  public float getValue(float parentValue) {
    return getValueInternal(mArray, mValue, mType, parentValue);
  }

  @Override
  public boolean equals(@Nullable Object obj) {
    if (obj == null) {
      return false;
    }
    PlatformLength other = (PlatformLength) obj;
    if (mType != other.mType) {
      return false;
    }
    if (mType != StyleConstants.PLATFORM_LENGTH_UNIT_CALC) {
      return FloatUtils.floatsEqual(mValue, other.mValue);
    }
    // FIXME: deep equals
    return mArray.asArrayList().equals(other.mArray.asArrayList());
  }

  public boolean isZero() {
    return mType != StyleConstants.PLATFORM_LENGTH_UNIT_CALC && mValue == 0;
  }

  private static float getValueInternal(
      ReadableArray arr, float value, int type, float parentValue) {
    if (type == StyleConstants.PLATFORM_LENGTH_UNIT_PERCENT) {
      return value * parentValue;
    } else if (type == StyleConstants.PLATFORM_LENGTH_UNIT_NUMBER) {
      return value;
    } else if (type == StyleConstants.PLATFORM_LENGTH_UNIT_CALC) {
      float ret = 0;
      for (int i = 0; i < arr.size(); i += 2) {
        ReadableArray item_array = null;
        float item_value = 0;
        int item_type = arr.getInt(i + 1);
        if (item_type == StyleConstants.PLATFORM_LENGTH_UNIT_CALC) {
          item_array = arr.getArray(i);
        } else if (item_type == StyleConstants.PLATFORM_LENGTH_UNIT_NUMBER
            || item_type == StyleConstants.PLATFORM_LENGTH_UNIT_PERCENT) {
          item_value = (float) arr.getDouble(i);
        }
        ret += getValueInternal(item_array, item_value, item_type, parentValue);
      }
      return ret;
    } else {
      return 0;
    }
  }
}

File: platform/harmony/lynx_harmony/src/main/cpp/ui/utils/platform_length.cc
// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

#include "platform/harmony/lynx_harmony/src/main/cpp/ui/utils/platform_length.h"

#include "base/include/value/array.h"

namespace lynx {
namespace tasm {
namespace harmony {

static float GetValueInternal(const fml::RefPtr<lepus::CArray>& calc_array,
                              float value, const PlatformLengthType unit,
                              float parent_value, float density) {
  if (unit == PlatformLengthType::kPercentage) {
    return value * parent_value;
  }
  if (unit == PlatformLengthType::kNumber) {
    return value * density;
  }
  if (unit == PlatformLengthType::kCalc) {
    float ret = 0;
    for (size_t i = 0; i < calc_array->size(); i += 2) {
      PlatformLengthType item_type =
          static_cast<PlatformLengthType>(calc_array->get(i + 1).Number());
      fml::RefPtr<lepus::CArray> item_array{nullptr};
      float item_value = 0;
      if (PlatformLengthType::kCalc == item_type) {
        item_array = calc_array->get(i).Array();
      } else if (item_type == PlatformLengthType::kNumber ||
                 item_type == PlatformLengthType::kPercentage) {
        item_value = calc_array->get(i).Number();
      }
      ret += GetValueInternal(item_array, item_value, item_type, parent_value,
                              density);
    }
    return ret;
  }
  return 0;
}

PlatformLength::PlatformLength() = default;

PlatformLength::PlatformLength(const lepus::Value& value,
                               const PlatformLengthType type) {
  type_ = type;
  if (type_ == PlatformLengthType::kCalc) {
    calc_array_ = value.Array();
  } else if (PlatformLengthType::kNumber == type ||
             PlatformLengthType::kPercentage == type) {
    val_ = value.Number();
  }
}

float PlatformLength::GetValue(float parent, float density) const {
  return GetValueInternal(calc_array_, val_, type_, parent, density);
}

}  // namespace harmony
}  // namespace tasm
}  // namespace lynx
